File: ue-sdk-ios/Sources/UESDK/TupleSnapshot.swift
/*
 * Silent Auth UE SDK (iOS) — Restlink (Ethiopia).
 * Device-side session-tuple snapshot. R&D only — never production.
 *
 * Copyright (c) 2026 Tran Nhan (nhanth87). All rights reserved.
 */

import Foundation

/// Point-in-time device tuple posted to SAS `POST /session-tuple`.
///
/// Devices cannot observe the real CGNAT source IP/port: `srcIp` and
/// `srcPort` are only ever set when the embedding app supplies them, so in
/// practice snapshots carry `ts` plus an optional app-supplied MSISDN.
/// The SDK itself never reads subscriber identifiers.
public struct TupleSnapshot: Codable, Equatable {

    public var srcIp: String?
    public var srcPort: Int?
    /// Capture timestamp, epoch milliseconds.
    public var ts: Int64
    public var msisdn: String?
    public var imsi: String?
    /// Bearer the tuple was captured on. Omitted from the wire when nil or
    /// unknown — the SAS treats an absent declaration as "not attested", never
    /// as cellular.
    public var accessTech: AccessTech?

    public init(srcIp: String? = nil,
                srcPort: Int? = nil,
                ts: Int64,
                msisdn: String? = nil,
                imsi: String? = nil,
                accessTech: AccessTech? = nil) {
        self.srcIp = srcIp
        self.srcPort = srcPort
        self.ts = ts
        self.msisdn = msisdn
        self.imsi = imsi
        self.accessTech = accessTech
    }

    /// Device-visible snapshot: capture timestamp now, no CGNAT visibility.
    public init(msisdn: String? = nil, accessTech: AccessTech? = nil) {
        let nowMillis = Int64(Date().timeIntervalSince1970 * 1000.0)
        self.init(ts: nowMillis, msisdn: msisdn, accessTech: accessTech)
    }

    private enum CodingKeys: String, CodingKey {
        case srcIp
        case srcPort
        case ts
        case msisdn
        case imsi
        case accessTech
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        srcIp = try container.decodeIfPresent(String.self, forKey: .srcIp)
        srcPort = try container.decodeIfPresent(Int.self, forKey: .srcPort)
        ts = try container.decode(Int64.self, forKey: .ts)
        msisdn = try container.decodeIfPresent(String.self, forKey: .msisdn)
        imsi = try container.decodeIfPresent(String.self, forKey: .imsi)
        accessTech = try container.decodeIfPresent(AccessTech.self, forKey: .accessTech)
    }

    /// Encodes only present fields so nulls are omitted from the wire body.
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(srcIp, forKey: .srcIp)
        try container.encodeIfPresent(srcPort, forKey: .srcPort)
        try container.encode(ts, forKey: .ts)
        try container.encodeIfPresent(msisdn, forKey: .msisdn)
        try container.encodeIfPresent(imsi, forKey: .imsi)
        if let accessTech = accessTech, accessTech != .unknown {
            try container.encode(accessTech, forKey: .accessTech)
        }
    }
}

File: ue-sdk-android/src/main/java/et/restlink/uesdk/android/TupleSnapshot.java
/*
 * Silent Auth UE SDK (Android) — Restlink (Ethiopia).
 * Device-side session-tuple poster. Java 8 / Android minSdk 24.
 * R&D only — never production.
 *
 * Copyright (c) 2026 Tran Nhan (nhanth87). All rights reserved.
 */

package et.restlink.uesdk.android;

/**
 * Point-in-time device tuple posted to SAS {@code /session-tuple}.
 *
 * <p>Devices cannot observe the real CGNAT source IP/port: {@code srcIp} and
 * {@code srcPort} are only ever set when the embedding app supplies them, so
 * in practice snapshots carry {@code ts}, an optional app-supplied MSISDN and
 * the {@link AccessTech} the SDK saw. The SDK itself never reads subscriber
 * identifiers.</p>
 *
 * <p>{@code accessTech} is not decoration: the SAS only trusts an IP&rarr;MSISDN
 * binding it can attribute to a cellular bearer, so a Wi-Fi tuple must be
 * identifiable as such rather than looking like a silent cellular one.</p>
 */
public final class TupleSnapshot {

    private final String srcIp;
    private final Integer srcPort;
    private final long ts;
    private final String claimedMsisdn;
    private final String imsi;
    private final AccessTech accessTech;

    public TupleSnapshot(String srcIp, Integer srcPort, long ts,
                         String claimedMsisdn, String imsi) {
        this(srcIp, srcPort, ts, claimedMsisdn, imsi, AccessTech.UNKNOWN);
    }

    public TupleSnapshot(String srcIp, Integer srcPort, long ts,
                         String claimedMsisdn, String imsi, AccessTech accessTech) {
        this.srcIp = srcIp;
        this.srcPort = srcPort;
        this.ts = ts;
        this.claimedMsisdn = claimedMsisdn;
        this.imsi = imsi;
        this.accessTech = accessTech == null ? AccessTech.UNKNOWN : accessTech;
    }

    /**
     * Device-visible snapshot: capture timestamp now, no CGNAT visibility.
     * Declares {@link AccessTech#UNKNOWN} — prefer {@link #cellularNow}.
     */
    public static TupleSnapshot now(String claimedMsisdnNullable) {
        return new TupleSnapshot(null, null, System.currentTimeMillis(),
                claimedMsisdnNullable, null);
    }

    /**
     * Snapshot taken on a {@link CellularBearer}: the declared access technology
     * travels with the tuple so the SAS can refuse a non-cellular binding.
     */
    public static TupleSnapshot cellularNow(String claimedMsisdnNullable, AccessTech accessTech) {
        return new TupleSnapshot(null, null, System.currentTimeMillis(),
                claimedMsisdnNullable, null, accessTech);
    }

    public String srcIp() {
        return srcIp;
    }

    public Integer srcPort() {
        return srcPort;
    }

    public long ts() {
        return ts;
    }

    public String claimedMsisdn() {
        return claimedMsisdn;
    }

    public String imsi() {
        return imsi;
    }

    /** Never null; {@link AccessTech#UNKNOWN} when the radio could not be read. */
    public AccessTech accessTech() {
        return accessTech;
    }
}

